pilot/pkg/kube/inject/inject.go

```diff
@@ -178,6 +179,13 @@ type Config struct {
 	// Template is the templated version of `SidecarInjectionSpec` prior to
 	// expansion over the `SidecarTemplateData`.
 	Template string `json:"template"`
+
+	// AlwaysInjectSelector: Forces the injection on pods whose labels match this selector.
+	AlwaysInjectSelector metav1.LabelSelector `json:"alwaysInjectSelector"`
+
+	// NeverInjectSelector: Refuses the injection on pods whose labels match this selector.
+	// Takes precedence over AlwaysInjectSelector.
+	NeverInjectSelector metav1.LabelSelector `json:"neverInjectSelector"`
```

Comment:
Do we need separate selectors for whitelist/blacklist? The LabelSelector's MatchExpressions support set-based options with non-existence semantics.

```yaml
# don't inject into pods with label `foo`
podSelector:
  matchExpression:
  - key: foo
    operator: DoesNotExist
- 

# only inject into pods with label `bar`
podSelector:
  matchExpression:
  - key: bar
    operator: Exist
- 
```

Reply:
I'm trying to be as more general as possible. Given multiple expressions are AND'ed, it might be difficult to express things like: Don't inject on these pods, but always inject those ones.

Also, having two selectors "AlwaysInject" and "NeverInject" seems to be more intuitive to use.

That said, I don't have a strong opinion on that; my use case is to add exceptions to the general rules (i.e., disabling injection on certain pods with a enabled namespace policy plus without touching the pod metadata/spec).

Reply:
Given that _deny_ has precedence over _allow_, I think the following addresses your first case:

```yaml
podSelector:
  matchExpression:
  - key: labelThatShouldBeNotBeInjected
    operator: DoesNotExist
  - key: labelThatShouldAlwaysBeInjected
    operator: Exist
```

Injection would be refused for any pod labeled with `labelThatShouldBeNotBeInjected`, regardless of other expressions. If that label isn't set, then `labelThatShouldAlwaysBeInjected` would take affect.

Also, we could start with a single selector and add a `denySelector` later with higher precedence if its really needed. It's harder to take configurations away..

Reply:
I'm not following...

In the simplest case, where I want to deny injection in a pod with the label `labelThatShouldBeNotBeInjected`:

```yaml
podSelector:
  matchExpression:
  - key: labelThatShouldBeNotBeInjected
    operator: DoesNotExist
```
This will be parsed and will **not** match. It's the same as if the pod doesn't have any label, and the processing would continue, defaulting to the enabled policy, which would inject...

Reply:
Annotation override should probably continue to have highest precedence over the selector, e.g. for backwards compatibility. The selector (allow/deny) override would apply if the selector was non-empty. Otherwise, we can fallback to the default global policy, e.g. 

```golang
if value, ok := annotations[sidecarAnnotationPolicyKey]; ok {
    // use annotation override if set
	switch strings.ToLower(value) {
 	case "y", "yes", "true", "on":	
        // inject
 	default:
        // don't inject
    }
} else if !selector.Empty() { 
    // use label selector if set
    if selector.Matches(labels.Set(metadata.Labels)) {
        // inject
    } else {
        // don't inject
    }
} else {    
    // otherwise use global policy
    if config.Policy == InjectionPolicyEnabled {
        // inject
    } else {
        // don't inject
    }
}
```

Reply:
Sorry for the delay, I'm (still) on vacation.

So, I don't think this approach would work well.
Supposing the policy is 'enabled' and I want to **not** inject on pods with a certain label, but leave injection enabled to others...

If I put:
```yaml
podSelector:
  matchExpression:
  - key: labelThatShouldBeNotBeInjected
    operator: Exists
```
It would match and pod would get injected.
If I put:
```yaml
podSelector:
  matchExpression:
  - key: labelThatShouldBeNotBeInjected
    operator: DoesNotExist
```
It would not match and the pod would not get injected, but that's the case for other pods that do not have that label. Causing all other pods not not get injected...

Am I missing something?